### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by Cindyzhang977**
Follow authentication tutorial to implement login feature.

- tell users to log in to leave a comment
- allow comments to be anonymous
  - toggle the required field for displayedName depending on whether "anonymous" is checked 
- display the user displayedName input or "anon" for each comment
- allow users to only delete the comments they made when logged in 
- display user account when logged in 
- implement log out button 
- add userEmail and displayedName attributes to each Comment object (server side)

image
image

### CODE REVIEW

File: portfolio/src/main/webapp/script.js [hunk 1]
```diff
@@ -274,13 +282,39 @@ function handleSubmit(e) {
     success: () => $('.comment').length > 0 ? loadComments(LoadType.RELOAD) : loadComments(LoadType.LOAD),
   });
   $(this).find('input,textarea').val('');
+  $("#anonCheck").prop("checked", false);
+  $('#displayedName').prop('required', true);
 }
 
 $('#rec-form').submit(handleSubmit);
+$('#anonCheck-container').click(() => {
+  $('#displayedName').prop('required', !$('#anonCheck').is(':checked'));
```
Comment (genedwards) [inline portfolio/src/main/webapp/script.js:291]: maybe we can keep the pattern of using prop
`$('#displayedName').prop('required', !$('#anonCheck').prop('checked'));`

File: portfolio/src/main/webapp/index.html [hunk 2]
```diff
@@ -153,6 +169,14 @@ <h4>Recommend A Destination</h4>
                 placeholder="Description (What to do, where specifically to go, general comments, etc.)"
               ></textarea>
             </div>
+            <div class="form-group col-auto">
+              <label class="sr-only" for="displayedName">Displayed Name</label>
+              <input type="text" class="form-control" id="displayedName" name="displayedName" aria-describedby="displayedName" placeholder="Displayed Name" required>
```
Comment (genedwards) [inline portfolio/src/main/webapp/index.html:174]: I'm a bit confused, aria-describedby has the id of its own element, is that intended?

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -190,7 +198,7 @@ function createComment(comment) {
               class=${linkClass}
               >Learn more</a
             >
-            <i class="fa fa-ban col-2" id="delete-btn-${comment.id}"></i>
+            <i class="fa fa-ban col-2" id="delete-btn-${comment.id}" style="display: ${deleteDisplay};"></i>
```
Comment (genedwards) [inline portfolio/src/main/webapp/script.js:201]: another option is to simply not render it

```
const deleteButton = $('#user-email').text() === comment.userEmail ?
  `<i class="fa fa-ban col-2" id="delete-btn-${comment.id}"></i>` : '';
return $(`
  ...
  ${deleteButton}
```
OR
```
const deleteButton = `<i class="fa fa-ban col-2" id="delete-btn-${comment.id}"></i>`;
return $(`
  ...
  ${$('#user-email').text() === comment.userEmail ? ${deleteButton} : ''}
```

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 4]
```diff
@@ -25,12 +25,16 @@
   private String link;
   private String description;
   private String id;
+  private String userEmail;
+  private String displayedName;
 
-  public Comment(String location, String link, @Nullable String description, String id) {
+  public Comment(String location, String link, @Nullable String description, String id, String userEmail, String displayedName) {
```
Comment (genedwards) [inline portfolio/src/main/java/com/google/sps/data/Comment.java:31]: ok for now, but if this keeps getting longer, refer to https://drive.google.com/file/d/1PJcAtx0N3fLJdJ9lLxeF5nALGuMxxy_7/view

File: portfolio/src/main/webapp/index.html (hunk 2 above)
Comment (Cindyzhang977) [reply portfolio/src/main/webapp/index.html:174]: Oops, I took this from bootstrap and I forgot to delete that part.

File: portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      this.isLoggedIn = isLoggedIn;
+      this.url = url;
+      this.userEmail = userEmail;
+    }
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    AuthStatus status;
+    UserService userService = UserServiceFactory.getUserService();
```
Comment (genedwards) [inline portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java:47]: I'm pretty sure Colm would ask this be `private final`

File: portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+      status = new AuthStatus(true, logoutUrl, userEmail);
+    } else {
+      String urlToRedirectToAfterUserLogsIn = "/";
+      String loginUrl = userService.createLoginURL(urlToRedirectToAfterUserLogsIn);
+      status = new AuthStatus(false, loginUrl, null);
+    }
+
+    response.setContentType("application/json;");
+    Gson gson = new Gson();
+    String json = gson.toJson(status);
```
Comment (genedwards) [inline portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java:61]: minor nit: If you're not keeping this gson around, you could single line it
```
String json = new Gson().toJson(status);
```
totally optional

**Pull request merged**